dspace-server-webapp/src/main/java/org/dspace/app/rest/SitemapRestController.java

```diff
@@ -0,0 +1,148 @@
+                    } else {
+                        throw new ResourceNotFoundException(
+                            "Directory with name " + name + " in " + sitemapOutputDir.getAbsolutePath() +
+                            " found, but no file.");
+                    }
+                }
+            }
+        } else {
+            throw new ResourceNotFoundException(
+                "Sitemap directory in " + sitemapOutputDir.getAbsolutePath() + " does not " +
+                "exist, either sitemaps have not been generated (./dspace generate-sitemaps)," +
+                " or are located elsewhere (config used: sitemap.dir).");
+        }
+        if (foundSitemapFile == null) {
+            throw new ResourceNotFoundException(
+                "Could not find sitemap file with name " + name + " in " + sitemapOutputDir.getAbsolutePath());
+        } else {
+            // return found sitemap file
+            this.returnSitemapFile(foundSitemapFile, response, request);
+        }
+    }
+
+    /**
+     * Sends back the matching sitemap file as a MultipartFile, with the headers set with details of the file
+     * (content, size, name, last modified)
+     *
+     * @param foundSitemapFile the found sitemap file, with matching name as in request path
+     * @param response         the HTTP response
+     * @param request          the HTTP request
+     * @throws SQLException if db error while completing DSpace context
+     * @throws IOException  if IO error surrounding sitemap file
+     */
+    private void returnSitemapFile(File foundSitemapFile, HttpServletResponse response, HttpServletRequest request)
+        throws SQLException, IOException {
+        // Pipe the bits
+        try (InputStream is = new FileInputStream(foundSitemapFile)) {
+            MultipartFileSender sender = MultipartFileSender
+                .fromInputStream(is)
+                .withBufferSize(BUFFER_SIZE)
+                .withFileName(foundSitemapFile.getName())
+                .withLength(foundSitemapFile.length())
+                .withMimetype(Files.probeContentType(foundSitemapFile.toPath()))
+                .with(request)
+                .with(response);
+
+            sender.withLastModified(foundSitemapFile.lastModified());
+
+            // Determine if we need to send the file as a download or if the browser can open it inline
+            long dispositionThreshold = configurationService.getLongProperty("webui.content_disposition_threshold");
+            if (dispositionThreshold >= 0 && foundSitemapFile.length() > dispositionThreshold) {
+                sender.withDisposition(MultipartFileSender.CONTENT_DISPOSITION_ATTACHMENT);
+            }
+
+            Context context = ContextUtil.obtainContext(request);
+
+            // We have all the data we need, close the connection to the database so that it doesn't stay open during
+            // download/streaming
+            context.complete();
```

Comment:
I think that @benbosman is right here, forcing the context to complete will free up resource earlier that we will never use. The context.complete method is already internally protected with an `isValid()` check we should eventually put a guard against NPE but this would be a very odd scenario so I'm also fine in just having the NPE and throw a 500

Reply:
Ok, sounds fine. I'll close this request.